### PR TITLE
New exercise: parser combinators.

### PR DESCRIPTION
**Pull request by fpottier**
This exercise builds on the nondeterminism monad, which has been implemented and used in several other exercises.

The code that must be written is not very complex, but the concepts in play can be difficult to grasp. Furthermore, there are several ungraded intermediate questions, and only 4 graded questions, so it is possible that some mistakes will not be immediately caught by the grading system. For these reasons, this exercise can be considered advanced.

### CODE REVIEW

File: exercises/parser_combinators/descr.md [hunk 1]
```diff
@@ -0,0 +1,636 @@
...
+```
+type 'a parser =
+  cursor -> ('a * cursor) NonDet.m
+```
+
+Thus, a parser is a function which, when applied to an initial cursor,
+constructs a nondeterministic computation which, when executed, produces a
+number of pairs of a result and a final cursor.
+
+Now, what is a cursor? It turns out that we need only one operation on
```
Comment (yurug) [inline exercises/parser_combinators/descr.md:274]: Don't you also need an operation to get the initial cursor for a given text?

File: exercises/parser_combinators/descr.md [hunk 2]
```diff
@@ -0,0 +1,636 @@
...
+
+## Parsing Arithmetic Expressions
+
+In the last part of this exercise, we would like to parse and evaluate
+arithmetic expressions. We begin with a very simple language where only
+integer literals, addition, and subtraction are permitted (Question 3).
+Then, we add multiplication, division, and the ability to enclose a
+subexpression within parentheses (Question 4).
+
+In the following, `'a top` is an abbreviation for the type of a binary
```
Comment (yurug) [inline exercises/parser_combinators/descr.md:479]:
```suggestion
In the following, `'a op` is an abbreviation for the type of a binary
```

File: exercises/parser_combinators/descr.md [hunk 3]
```diff
@@ -0,0 +1,636 @@
...
+
+**Question.** Define a parser `multiplicative_op` of type `int op
+parser` that recognizes one of the characters `'*'` or `'/'` and
+interprets it as the corresponding operation on integers.
+
+**Question.** Define a parser combinator `chainl1` of type
+`'a parser -> 'a op parser -> 'a parser` such that the parser
+`chainl1 p op` recognizes a nonempty sequence of `p`'s,
+separated with `op`'s,
+and interprets it as a left-associative chain of applications
```
Comment (yurug) [inline exercises/parser_combinators/descr.md:503]:
```suggestion
and interprets it as a left-parenthesized chain of applications
```

File: exercises/parser_combinators/descr.md [hunk 4]
```diff
@@ -0,0 +1,636 @@
...
+must be consumed.
+
+**Question 4.** Define a parser `term` of type `int parser` that recognizes
+and evaluates a **term**, as defined by the above grammar. For instance, the
+string `"49-12*(2+2)"` should be evaluated to the single integer result `1`.
+
+*Hint.* One approach is to define `atom`, `factor` and `term` as three
+mutually recursive functions. An alternative approach is to use the parser
+combinator `fix`, whose type is `('a parser -> 'a parser) -> 'a parser`,
+and whose definition is given to you.
```
Comment (yurug) [inline exercises/parser_combinators/descr.md:561]: in the prelude, I guess?

File: exercises/parser_combinators/descr.md (hunk 1 above)
Comment (fpottier) [reply exercises/parser_combinators/descr.md:274]: Indeed, this is explained later on, where I write: "the input string `input` must be converted to a sequence of characters, therefore, a cursor".

File: exercises/parser_combinators/descr.md (hunk 4 above)
Comment (fpottier) [reply exercises/parser_combinators/descr.md:561]: At the moment, it is not in the prelude; it is part of the template.

**Comment by yurug** [comment]
Thanks!

**Pull request merged**